core/src/main/java/org/apache/gravitino/EntityStoreV2.java

```diff
@@ -0,0 +1,219 @@
+   * retrieve of entities.
+   *
+   * @param ident the unique identifier of the entity
+   * @param entityType the general type of the entity
+   * @param e the entity class instance
+   * @param <E> the class of entity
+   * @return the entity retrieved from the underlying storage
+   * @throws NoSuchEntityException if the entity does not exist
+   * @throws IOException if the retrieve operation fails
+   */
+  <E extends Entity & HasIdentifier> E get(NameIdentifier ident, EntityType entityType, Class<E> e)
+      throws NoSuchEntityException, IOException;
+
+  /**
+   * Delete the entity from the underlying storage by the specified {@link
+   * org.apache.gravitino.NameIdentifier}.
+   *
+   * @param ident the name identifier of the entity
+   * @param entityType the type of the entity to be deleted
+   * @return true if the entity exists and is deleted successfully, false otherwise
+   * @throws IOException if the delete operation fails
+   */
+  default boolean delete(NameIdentifier ident, EntityType entityType) throws IOException {
+    return delete(ident, entityType, false);
+  }
+
+  /**
+   * Delete the entity from the underlying storage by the specified {@link
+   * org.apache.gravitino.NameIdentifier}.
+   *
+   * @param ident the name identifier of the entity
+   * @param entityType the type of the entity to be deleted
+   * @param cascade support cascade delete or not
+   * @return true if the entity exists and is deleted successfully, false otherwise
+   * @throws IOException if the delete operation fails
+   */
+  boolean delete(NameIdentifier ident, EntityType entityType, boolean cascade) throws IOException;
+
+  /**
+   * List all the entities with the specified {@link org.apache.gravitino.NameIdentifier}, and
+   * deserialize them into the specified {@link Entity} object.
+   *
+   * <p>Note. Depends on the isolation levels provided by the underlying storage, the returned list
+   * may not be consistent.
+   *
+   * @param ident the name identifier of the entity
+   * @param <E> class of the entity
+   * @param type the detailed type of the entity
+   * @param entityType the general type of the entity
+   * @param relationType the relation type of entities to be list out with the specified entity
+   * @throws IOException if the list operation fails
+   * @return the list of entities
+   */
+  <E extends Entity & HasIdentifier> List<E> listEntitiesByRelation(
+      NameIdentifier ident, Class<E> type, EntityType entityType, RelationType relationType)
+      throws IOException;
+
+  /**
+   * @param start NameIdentifier of the starting point of the relationship.
+   * @param type EntityType of the starting point.
+   * @param relationType RelationType of the relationship.
+   * @return List of MetadataObject that are related to the starting point.
+   * @throws IOException if the list operation fails.
+   */
+  List<MetadataObject> listMetadataObjectsByRelation(
```

Comment:
What's the difference between `listEntitiesByRelation` and `listMetadataObjectsByRelation`?

Reply:
They are utilized in different scenarios. 

- `listEntitiesByRelation` is used to list all entities by a creation, all these entities should be with the same type, for example, there can be all `metalake` or `catalogs`, currently we can't get a bundle of entities with different entity types.
- `listMetadataObjectsByRelation` handles the case that `listEntitiesByRelation` can't cope with and only returns entities identifier. 

Do you have any good suggestions for merging these two, I'm also trying to combine them, the current solution is not so elegant.

Reply:
@jerqi 
Currently the concepts of `Entity` and `MetadataObject` are quite different, and I can hardly find a good way to merge them.

Reply:
So we directly load the entity and then convert it to `MetadataObject`? If so, it's very heavy when loading/listing metadataobjects.

Reply:
It's ok for me. The convert is the memory operation.

Reply:
Okay, let me review it again.